client/rpc/src/eth/block.rs

```diff
@@ -93,44 +96,90 @@ where
 			client.as_ref(),
 			backend.as_ref(),
 			Some(number),
 		)? {
-			Some(id) => id,
-			None => return Ok(None),
-		};
-		let substrate_hash = client
-			.expect_block_hash_from_id(&id)
-			.map_err(|_| internal_err(format!("Expect block number from id: {}", id)))?;
+			Some(id) => {
+				let substrate_hash = client
+					.expect_block_hash_from_id(&id)
+					.map_err(|_| internal_err(format!("Expect block number from id: {}", id)))?;
 
-		let schema = fc_storage::onchain_storage_schema(client.as_ref(), substrate_hash);
+				let schema = fc_storage::onchain_storage_schema(client.as_ref(), substrate_hash);
 
-		let block = block_data_cache.current_block(schema, substrate_hash).await;
-		let statuses = block_data_cache
-			.current_transaction_statuses(schema, substrate_hash)
-			.await;
+				let block = block_data_cache.current_block(schema, substrate_hash).await;
+				let statuses = block_data_cache
+					.current_transaction_statuses(schema, substrate_hash)
+					.await;
 
-		let base_fee = client
-			.runtime_api()
-			.gas_price(substrate_hash)
-			.unwrap_or_default();
+				let base_fee = client
+					.runtime_api()
+					.gas_price(substrate_hash)
+					.unwrap_or_default();
 
-		match (block, statuses) {
-			(Some(block), Some(statuses)) => {
-				let hash = H256::from(keccak_256(&rlp::encode(&block.header)));
-
-				Ok(Some(rich_block_build(
-					block,
-					statuses.into_iter().map(Option::Some).collect(),
-					Some(hash),
-					full,
-					Some(base_fee),
-				)))
+				match (block, statuses) {
+					(Some(block), Some(statuses)) => {
+						let hash = H256::from(keccak_256(&rlp::encode(&block.header)));
+
+						Ok(Some(rich_block_build(
+							block,
+							statuses.into_iter().map(Option::Some).collect(),
+							Some(hash),
+							full,
+							Some(base_fee),
+							false,
+						)))
+					}
+					_ => Ok(None),
+				}
 			}
-			_ => Ok(None),
+			None if number == BlockNumber::Pending => {
+				let api = client.runtime_api();
+				let best_hash = client.info().best_hash;
+
+				// Get current in-pool transactions
+				let mut xts: Vec<<B as BlockT>::Extrinsic> = Vec::new();
+				// ready validated pool
+				xts.extend(
+					graph
+						.validated_pool()
+						.ready()
+						.map(|in_pool_tx| in_pool_tx.data().clone())
+						.collect::<Vec<<B as BlockT>::Extrinsic>>(),
+				);
+
+				// future validated pool
+				xts.extend(
+					graph
+						.validated_pool()
+						.futures()
+						.iter()
+						.map(|(_hash, extrinsic)| extrinsic.clone())
+						.collect::<Vec<<B as BlockT>::Extrinsic>>(),
+				);
+
+				let (block, statuses) = api
+					.pending_block(best_hash, xts)
+					.map_err(|_| internal_err(format!("Runtime access error at {}", best_hash)))?;
+
+				let base_fee = api.gas_price(best_hash).unwrap_or_default();
+
+				match (block, statuses) {
+					(Some(block), Some(statuses)) => Ok(Some(rich_block_build(
+						block,
+						statuses.into_iter().map(Option::Some).collect(),
+						None,
```

Comment:
The `hash` field of result should be `None` if we support the `pending` tag, but the function `rich_block_build` will give the default value for the `hash` field if it's `None`.

```rust
header: Header {
    hash: Some(
	    hash.unwrap_or_else(|| H256::from(keccak_256(&rlp::encode(&block.header)))),
	),
	parent_hash: block.header.parent_hash,
    ...
},
```